### ISSUE TITLE
.function_indexes does not detect all functions

### ISSUE THREAD

**Issue by vStone** [puppetlabs/puppet-lint#230]
`.function_indexes` only supports a very limited set of functions in specific cases.
It should contain all possible functions in a manifest.

Example code:

```
class foobar {
  $x = function_assign_to_var('foobar')
  $y = 'foobar'.function_assign_inline_no_parenthesis
  $z = 'foobar'.function_assign_inline_with_args('arg2')
  function_call_no_args
  function_call_with_args('xxx') # detected
  function_call_parent(function_call_nested('x')) # both functions are detected.
  function_call_nested2.with_inline_function
  chain.multiple().calls('to').functions
}
```

.function_indexes only contains 3 functions (see comments)

It should contain
* `function_assign_to_var`
* `function_assign_inline_no_parenthesis`
* `function_assign_inline_with_args`
* `function_call_no_args`
* `function_call_with_args`
* `function_call_parent`
* `function_call_nested`
* `function_call_nested2`
* `with_inline_function`
* `chain`
* `multiple`
* `calls`
* `functions`

It also seems that there are no tests for function_indexes.
